quickwit/quickwit-actors/src/scheduler.rs

```diff
@@ -58,139 +52,196 @@ impl PartialOrd for TimeoutEvent {
-        &mut self,
-        message: SimulateAdvanceTime,
-        ctx: &ActorContext<Self>,
-    ) -> Result<(), ActorExitStatus> {
-        self.process_simulate_advance_time(message.time_shift, message.tx, ctx)
-            .await;
-        Ok(())
+impl NoAdvanceTimeGuard {
+    fn new(scheduler_client: SchedulerClient) -> Self {
+        scheduler_client.inc_no_advance_time();
+        NoAdvanceTimeGuard { scheduler_client }
     }
 }
 
-#[async_trait]
-impl Handler<ScheduleEvent> for Scheduler {
-    type Reply = ();
+impl Drop for NoAdvanceTimeGuard {
+    fn drop(&mut self) {
+        self.scheduler_client.dec_no_advance_time();
+    }
+}
 
-    async fn handle(
-        &mut self,
-        message: ScheduleEvent,
-        ctx: &ActorContext<Self>,
-    ) -> Result<(), ActorExitStatus> {
-        self.process_schedule_event(message.timeout, message.callback, ctx)
-            .await;
-        Ok(())
+struct AccelerateTimeCountGuard(SchedulerClient);
+
+impl AccelerateTimeCountGuard {
+    fn new(scheduler_client: SchedulerClient) -> Self {
+        scheduler_client
+            .inner
+            .accelerate_time_count
+            .fetch_add(1, Ordering::SeqCst);
+        AccelerateTimeCountGuard(scheduler_client)
     }
 }
 
-#[async_trait]
-impl Handler<Timeout> for Scheduler {
-    type Reply = ();
-
-    async fn handle(
-        &mut self,
-        _message: Timeout,
-        ctx: &ActorContext<Self>,
-    ) -> Result<(), ActorExitStatus> {
-        self.process_timeout(ctx).await;
-        Ok(())
+impl Drop for AccelerateTimeCountGuard {
+    fn drop(&mut self) {
+        self.0
+            .inner
+            .accelerate_time_count
+            .fetch_sub(1, Ordering::SeqCst);
     }
 }
 
+pub fn start_scheduler() -> SchedulerClient {
+    let (tx, rx) = flume::unbounded::<SchedulerMessage>();
+    let scheduler_client = SchedulerClient {
+        inner: Arc::new(SchedulerClientInner {
+            no_advance_time_guard_count: AtomicUsize::default(),
+            accelerate_time_count: AtomicUsize::default(),
+            tx,
+        }),
+    };
+    let mut scheduler = Scheduler::new(&scheduler_client);
+    tokio::spawn(async move {
+        while let Ok(scheduler_message) = rx.recv_async().await {
+            match scheduler_message {
+                SchedulerMessage::Timeout => scheduler.process_timeout().await,
+                SchedulerMessage::Schedule { callback, timeout } => {
+                    scheduler.process_schedule(callback, timeout);
+                }
+                SchedulerMessage::SimulateSleep { cb, duration } => {
+                    scheduler.process_simulate_sleep(cb, duration).await;
+                }
+            }
+        }
+    });
+    scheduler_client
+}
+
+struct Scheduler {
+    // We attribute an event_id to all event just to break ties
+    // if two events are scheduled on the same time.
+    event_id_generator: u64,
+    // simulated_time = Instant::now() + simulated_time_shift.
```

Comment:
nitpick: docs

```rust
    // Simulated time shift which defines the scheduler time reference as `simulated_time = Instant::now() + simulated_time_shift`.
    // By default `simulated_time_shift` is set to 0 but it can be shifted when the scheduler has to process a simulate sleep event`.
```